xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/ClusterManager.java

```diff
@@ -0,0 +1,272 @@
+
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Streams;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.CommonPools;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.AsyncCloseable;
+import com.linecorp.armeria.common.util.Listenable;
+import com.linecorp.armeria.common.util.SafeCloseable;
+import com.linecorp.armeria.internal.common.util.ReentrantShortLock;
+import com.linecorp.armeria.xds.ClusterSnapshot;
+import com.linecorp.armeria.xds.ListenerRoot;
+import com.linecorp.armeria.xds.ListenerSnapshot;
+import com.linecorp.armeria.xds.RouteSnapshot;
+import com.linecorp.armeria.xds.SnapshotWatcher;
+import com.linecorp.armeria.xds.client.endpoint.ClusterManager.State;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class ClusterManager implements SnapshotWatcher<ListenerSnapshot>, AsyncCloseable,
+                                      Listenable<State>, Consumer<List<Endpoint>> {
+
+    private static final Logger logger = LoggerFactory.getLogger(ClusterManager.class);
+    private static final Map<ClusterSnapshot, ClusterEntry> INITIAL_CLUSTER_ENTRIES = new HashMap<>();
+
+    private final EventExecutor eventLoop;
+    private final SafeCloseable safeCloseable;
+    private final CompletableFuture<Void> closeFuture = new CompletableFuture<>();
+
+    private volatile Map<ClusterSnapshot, ClusterEntry> clusterEntries = INITIAL_CLUSTER_ENTRIES;
+    @Nullable
+    private volatile ListenerSnapshot listenerSnapshot;
+    private final Set<CompletableFuture<?>> pendingRemovals = Sets.newConcurrentHashSet();
+    private boolean closed;
+
+    @GuardedBy("listenersLock")
+    private final List<Consumer<? super State>> listeners = new ArrayList<>();
+    private final ReentrantShortLock listenersLock = new ReentrantShortLock();
+
+    ClusterManager(ListenerRoot listenerRoot) {
+        eventLoop = listenerRoot.eventLoop();
+        safeCloseable = listenerRoot;
+        listenerRoot.addSnapshotWatcher(this);
+    }
+
+    ClusterManager(ClusterSnapshot clusterSnapshot) {
+        checkArgument(clusterSnapshot.endpointSnapshot() != null,
+                      "An endpoint snapshot must exist for the provided (%s)", clusterSnapshot);
+        eventLoop = CommonPools.workerGroup().next();
+        clusterEntries = ImmutableMap.of(clusterSnapshot, new ClusterEntry(clusterSnapshot, this));
+        safeCloseable = () -> {};
+    }
+
+    @Nullable
+    Endpoint selectNow(ClientRequestContext ctx) {
+        final Map<ClusterSnapshot, ClusterEntry> clusterEntries = this.clusterEntries;
+        for (Entry<ClusterSnapshot, ClusterEntry> entry: clusterEntries.entrySet()) {
```

Comment:
nit:
```suggestion
        for (Entry<ClusterSnapshot, ClusterEntry> entry : clusterEntries.entrySet()) {
```
I also found several same mistakes in this class.